File: app/src/main/java/com/warchaser/musicplayer/main/SongsAdapter.java
package com.warchaser.musicplayer.main;

import android.content.Context;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import com.bumptech.glide.Glide;
import com.warchaser.musicplayer.R;
import com.warchaser.musicplayer.tools.FormatHelper;
import com.warchaser.musicplayer.tools.MusicInfo;
import com.warchaser.musicplayer.tools.MusicList;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Wucn on 2017/12/20.
 * Songs Adapter
 */
public class SongsAdapter extends RecyclerView.Adapter<SongsAdapter.ViewHolderItem> {

    private Context mContext;
    private OnItemClickListener mOnItemClickListener;

    private OnItemClickDelegate mOnItemClickDelegate;

    private int mCurrentPosition = MusicList.getCurrentMusicInt();

    SongsAdapter(Context context) {
        mContext = context;
        mOnItemClickListener = new OnItemClickListener();
    }

    @NonNull
    @Override
    public ViewHolderItem onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewHolderItem(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderItem holder, int position, @NonNull List<Object> payloads) {
        boolean isFullRefresh = payloads.isEmpty();

        if(isFullRefresh){
            MusicInfo bean = MusicList.getMusicWithPosition(position);

            Glide.with(mContext).asGif().load(R.mipmap.moving_music).into(holder.gfGo);

            holder.tvItemTitle.setText(bean.getTitle());
            holder.tvItemDuration.setText(FormatHelper.formatDuration(bean.getDuration()));

            holder.tvItemTitle.setTextColor(Color.argb(255, 0, 0, 0));
            holder.tvItemDuration.setTextColor(Color.argb(255, 0, 0, 0));

            holder.gfGo.setVisibility(View.INVISIBLE);

            holder.mLyRoot.setTag(position);
            holder.mLyRoot.setOnClickListener(mOnItemClickListener);

            holder.mBtnMenu.setTag(position);
            holder.mBtnMenu.setOnClickListener(mOnItemClickListener);
        }

        if (position == MusicList.getCurrentMusicInt()) {
            holder.tvItemTitle.setTextColor(Color.RED);
            holder.tvItemDuration.setTextColor(Color.RED);
            holder.gfGo.setVisibility(View.VISIBLE);
        } else {
            holder.tvItemTitle.setTextColor(Color.argb(255, 0, 0, 0));
            holder.tvItemDuration.setTextColor(Color.argb(255, 0, 0, 0));

            holder.gfGo.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderItem viewHolderItem, int position) {
        onBindViewHolder(viewHolderItem, position, new ArrayList<>());
    }

    @Override
    public int getItemCount() {
        return MusicList.size();
    }

    /**
     * 配合setHasStableIds(true)使用
     * */
    @Override
    public long getItemId(int position) {
        return position;
    }

    public void setOnItemClickDelegate(OnItemClickDelegate delegate){
        this.mOnItemClickDelegate = delegate;
    }

    private class OnItemClickListener implements View.OnClickListener{

        @Override
        public void onClick(View v) {

            if(mOnItemClickDelegate == null){
                return;
            }

            final int position = (int)v.getTag();
            final MusicInfo bean = MusicList.getMusicWithPosition(position);
            switch (v.getId()){
                case R.id.mLyRoot:
                    notifyItemsChanged(position);
                    mOnItemClickDelegate.onItemClick(position, bean);
                    break;
                case R.id.mBtnMenu:
                    mOnItemClickDelegate.onMenuClick(position, bean);
                    break;
                default:
                    break;
            }
        }
    }

    public void notifyItemsChanged(int position){
        notifyItemChanged(mCurrentPosition, "local_refresh");
        notifyItemChanged(position, "local_refresh");
        mCurrentPosition = position;
    }

    public interface OnItemClickDelegate{

        void onItemClick(int position, MusicInfo bean);

        void onMenuClick(int position, MusicInfo bean);
    }

    class ViewHolderItem extends RecyclerView.ViewHolder{
        /**
         * TextView, Music Title
         */
        @BindView(R.id.tvItemTitle)
        TextView tvItemTitle;

        /**
         * TextView, Music Duration
         */
        @BindView(R.id.tvItemDuration)
        TextView tvItemDuration;

        /**
         * GifView, Tag a Gif on Current Music which is Playing
         */
        @BindView(R.id.gfGo)
        ImageView gfGo;

        /**
         * 右侧菜单按钮
         * */
        @BindView(R.id.mBtnMenu)
        Button mBtnMenu;

        @BindView(R.id.mLyRoot)
        RelativeLayout mLyRoot;

        ViewHolderItem(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/com/warchaser/musicplayer/main/SongsAdapter.kt
package com.warchaser.musicplayer.main

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.warchaser.musicplayer.R
import com.warchaser.musicplayer.tools.FormatHelper
import com.warchaser.musicplayer.tools.MusicInfo
import com.warchaser.musicplayer.tools.MusicList
import kotlinx.android.synthetic.main.item.view.*

class SongsAdapter(context: Context) : RecyclerView.Adapter<SongsAdapter.ViewHolderItem>() {

    private val mContext : Context = context

    private var mOnItemClickDelegate : OnItemClickDelegate? = null

    private var mCurrentPosition : Int = MusicList.getCurrentMusicInt()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderItem {
        val rootView = LayoutInflater.from(parent.context).inflate(R.layout.item, parent, false)
        return ViewHolderItem(rootView).apply {
            val clickListener = OnItemClickListener(this)
            rootView.run {
                setOnClickListener(clickListener)
                findViewById<RelativeLayout>(R.id.mLyRoot).setOnClickListener(clickListener)
                findViewById<Button>(R.id.mBtnMenu).setOnClickListener(clickListener)
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolderItem, position: Int, payloads: MutableList<Any>) {
        val isFullRefresh = payloads.isEmpty()

        holder.itemView.apply {
            if(isFullRefresh){
                val bean = MusicList.getMusicWithPosition(position)
                Glide.with(mContext).asGif().load(R.mipmap.moving_music).into(mIvGo)

                mTvTitle.text = bean.title
                mTvDuration.text = FormatHelper.formatDuration(bean.duration)

                mTvTitle.setTextColor(Color.BLACK)
                mTvDuration.setTextColor(Color.BLACK)

                mIvGo.visibility = View.INVISIBLE
            }

            if(position == MusicList.getCurrentMusicInt()){
                mTvTitle.setTextColor(Color.RED)
                mTvDuration.setTextColor(Color.RED)
                mIvGo.visibility = View.VISIBLE
            } else {
                mTvTitle.setTextColor(Color.BLACK)
                mTvDuration.setTextColor(Color.BLACK)
                mIvGo.visibility = View.INVISIBLE
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolderItem, position: Int) {
        onBindViewHolder(holder, position, ArrayList())
    }

    override fun getItemCount(): Int {
        return MusicList.size()
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    fun setOnItemClickDelegate(delegate: OnItemClickDelegate){
        mOnItemClickDelegate = delegate
    }

    fun notifyItemsChanged(position: Int){
        notifyItemChanged(mCurrentPosition, "local_refresh")
        notifyItemChanged(position, "local_refresh")
        mCurrentPosition = position
    }

    private inner class OnItemClickListener(val mHolder : RecyclerView.ViewHolder) : View.OnClickListener{

        override fun onClick(v: View?) {

            mOnItemClickDelegate?.run {
                val position = mHolder.adapterPosition
                if(position == RecyclerView.NO_POSITION){
                    return
                }
                val bean = MusicList.getMusicWithPosition(position)
                when(v!!.id){
                    R.id.mLyRoot -> {
                        notifyItemsChanged(position)
                        onItemClick(position, bean)
                    }
                    R.id.mBtnMenu -> {
                        onMenuClick(position, bean)
                    }
                }
            }
        }
    }

    interface OnItemClickDelegate{
        fun onItemClick(position: Int, bean : MusicInfo)

        fun onMenuClick(position: Int, bean : MusicInfo)
    }

    class ViewHolderItem(view : View) : RecyclerView.ViewHolder(view)
}